docs/dev/hacking-guide.md

```diff
@@ -0,0 +1,233 @@
+
+https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md#machine-lifecycle-diagram
+
+## Table of contents
+- [I'm completely new here, where to start?](#im-completely-new-here-where-to-start?)
+  * [General information about Machine API structure](#general-information-about-machine-api-structure)
+  * [How to start contributing](#how-to-start-contributing)
+- [How to run unit tests](#how-to-run-unit-tests)
+- [How to run a component locally for testing](#how-to-run-a-component-locally-for-testing)
+   * [Running machine controller](#running-machine-controller)
+- [How to build the software in a container for remote testing](#how-to-build-the-software-in-a-container-for-remote-testing)
+- [How to run e2e tests](#how-to-run-e2e-tests)
+  * [Running specific e2e tests](#running-specific-e2e-tests)
+- [How to update dependencies](#how-to-update-dependencies)
+- [How to update generated artifacts](#how-to-update-generated-artifacts)
+- [How to use something other than Docker to run make targets](#how-to-use-something-other-than-Docker-to-run-make-targets)
+- [Some links to the CI configuration](#some-links-to-the-ci-configuration)
+- [Where to file an issue](#where-to-file-an-issue)
+
+## I'm completely new here, where to start?
+### General information about Machine API structure
+Machine API consists of a couple different components:
+
+- Machine API Operator - https://github.com/openshift/machine-api-operator. The operator contains CRDs(Machine,MachineSet, MachineHeathCheck) and is responsible for running controllers for given CRDs. It’s also responsible for running the controllers on the cluster.
+
+- Provider specific implementation. Each provider except vSphere is located in a separate repository. These providers use the code from the machine controller(link to MAO/machine-controller) and extend it with cloud specific logic:
+  - https://github.com/openshift/cluster-api-provider-gcp
+  - https://github.com/openshift/cluster-api-provider-azure
+  - https://github.com/openshift/cluster-api-provider-aws
+  - https://github.com/openshift/machine-api-operator/tree/master/pkg/controller/vsphere
+  - https://github.com/openshift/cluster-api-provider-openstack
+  - https://github.com/openshift/cluster-api-provider-baremetal
+  - https://github.com/openshift/cluster-api-provider-ovirt
+
+### How to start contributing
+Prerequisites:
+- Running OCP cluster
+- Installed Golang
+
+First you need to determine if the change is related to cloud provider specific logic or it's related to all providers? Maybe you want to contribute to our e2e test suite?
+
+If that's a cloud provider related change, you should fork and clone one of provider repositories, the list can be found above.
+
+If the change is related to all providers, you should fork and clone machine-api-operator.
+
+In order to add changes to our e2e test suite fork and clone this repository https://github.com/openshift/cluster-api-actuator-pkg.
```

Comment:
Wonder if we should link these so that are more readable? WDYT?
```suggestion
In order to add changes to our e2e test suite fork and clone [cluster-api-actuator-pkg](https://github.com/openshift/cluster-api-actuator-pkg).
```